rubix-spi/src/main/java/com/qubole/rubix/spi/fop/ObjectPoolPartition.java

```diff
@@ -182,4 +185,55 @@ public synchronized int getTotalCount()
   {
     return totalCount;
   }
+
+  // set the scavenge interval carefully
+  public synchronized void scavenge() throws InterruptedException
+  {
+    int delta = this.totalCount - config.getMinSize();
+    if (delta <= 0) {
+      log.debug(this.name + " : Scavenge for delta <= 0, Skipping !!!");
+      return;
+    }
+    int removed = 0;
+    long now = System.currentTimeMillis();
+    Poolable<T> obj;
+    obj = objectQueue.poll();
+    while (delta-- > 0 && obj != null) {
+      // performance trade off: delta always decrease even if the queue is empty,
+      // so it could take several intervals to shrink the pool to the configured min value.
+      log.debug(String.format("%s : obj=%s, now-last=%s, max idle=%s", this.name, obj, now - obj.getLastAccessTs(),
+              config.getMaxIdleMilliseconds()));
+      if (now - obj.getLastAccessTs() > config.getMaxIdleMilliseconds() &&
+              ThreadLocalRandom.current().nextDouble(1) < config.getScavengeRatio()) {
+        log.debug(this.name + " : Scavenger removing object: " + obj);
+        decreaseObject(obj); // shrink the pool size if the object reaches max idle time
+        removed++;
+      }
+      else {
+        objectQueue.put(obj); //put it back
```

Comment:
we can break the current cycle here because objects on objectQueue are in order of their accessTime (roughly). Even though there can be exceptional cases of an object 'o1' with older access time appearing after an object 'o2' of more recent access time due to 'o1' being borrowed and used for a longer period but i think such cases are fine because we do not know the real accessTime between the time of borrowing and the time of return.

Reply:
yeah, we can do that to save some time.

Reply:
fixed.